File: 2021/day4/day4.hs
import Data.List

replaceCommas :: String -> String
replaceCommas str =
    let
        repl ',' = ' '
        repl c = c
    in
        map repl str

parseLine :: String -> [Int]
parseLine line = map (\n -> read n :: Int) (words line)

parseBoard :: [String] -> [[Int]]
parseBoard = map parseLine

-- Boards are 5 lines tall, with an empty line between them
parseBoards :: [String] -> [[[Int]]]
parseBoards [] = []
parseBoards ls = parseBoard (take 5 ls):parseBoards (drop 6 ls)

scoreBoard :: [Int] -> [[Int]] -> Int
scoreBoard draws board =
    last draws * sum (filter (`notElem` draws) (concat board))

lineWins :: [Int] -> [Int] -> Bool
lineWins draws = all (`elem` draws)

boardWins :: [Int] -> [[Int]] -> Bool
boardWins draws board =
    any (lineWins draws) board || any (lineWins draws) (transpose board)

part1 :: [Int] -> Int -> [[[Int]]] -> Int
part1 draws i boards
    | null winningBoards = part1 draws (i + 1) boards
    | otherwise = scoreBoard upTo (head winningBoards) 
    where
        upTo = take i draws
        winningBoards = filter (boardWins upTo) boards

part2 :: [Int] -> Int -> [[[Int]]] -> Int
part2 draws i [b]
    | boardWins upTo b = scoreBoard (take i draws) b
    | otherwise = part2 draws (i + 1) [b]
    where upTo = take i draws
part2 draws i boards =
    let
        upTo = take i draws
    in
        part2 draws (i + 1) $ filter (not . boardWins upTo) boards

main :: IO ()
main = do
    drawsLine <- getLine
    let draws = parseLine $ replaceCommas drawsLine
    getLine
    boardsLines <- getContents
    let boards = parseBoards $ lines boardsLines
    putStrLn $ "Part 1: " ++ show (part1 draws 1 boards)
    putStrLn $ "Part 2: " ++ show (part2 draws 1 boards)

File: 2022/day4.py
with open('day4.txt', 'r') as f:
    data = f.read().strip().split('\n')

part1 = 0
part2 = 0

for pair in data:
    a, b = pair.split(',')
    a1, a2 = map(int, a.split('-'))
    b1, b2 = map(int, b.split('-'))
    if (a1 <= b1 and a2 >= b2) or (b1 <= a1 and b2 >= a2):
        part1 += 1
        part2 += 1
    elif (a1 <= b1 and a2 >= b1) or (a1 <= b2 and a2 >= b2):
        part2 += 1

print('Part 1:', part1, 'Ranges')
print('Part 2:', part2, 'Ranges')

